lib/ui/painting/gradient.cc

```diff
@@ -46,14 +46,17 @@ void CanvasGradient::initLinear(const tonic::Float32List& end_points,
   SkPoint p0 = SkPoint::Make(end_points[0], end_points[1]);
   SkPoint p1 = SkPoint::Make(end_points[2], end_points[3]);
   std::vector<DlColor> dl_colors;
-  dl_colors.reserve(colors.num_elements());
-  for (int i = 0; i < colors.num_elements(); ++i) {
-    /// TODO(gaaclarke): Make this preserve wide gamut colors.
-    dl_colors.emplace_back(DlColor(colors[i]));
+  dl_colors.reserve(color_stops.num_elements());
```

Comment:
Can't this go back to a cast now? We're copying the data into the display list again so we shouldn't need to allocate more storage.

Reply:
Oh, we can't cast it anymore since the color space field.

In that case, can we update this API so we reserve and write the color bytes directly into the DL storage instead of allocating an intermediary?

Reply:
In order to have a cast the memory layouts need to be identical.  Here we have an array of floats, we need to insert the colorspace field into the memory layout if we wanted to use casts.  The casts are super frail.  I don't recommend it unless we get some data that supports it's worth the risk.

If we did find it was slowing us down, I would recommend doing the conversion on the stack when the number of colors is less than a certain size to catch most use cases before resorting to an unsafe cast.

Reply:
Something like this:
```c++
if (num_colors <= 3) {
  std::array<DlColor, 3> dl_colors;
  //...
} else {
  std::vector<DlColor> dl_colors;
  dl_colors.reserve(num_colors);
  //...
}
```

I don't think it's worth the extra complexity, but I can do it if we want.  I know we were bit in the past with too many heap allocations.

Reply:
All this code is doing is passing the data into the DLBuilder so it can be copied into the display list. Can we copy into the final destination instead of allocating an intermediary?

Reply:
Don't we need to copy the values somehow to isolate them from changes in the app? If so, then it doesn't matter if we convert before or after the copy. Also, this conversion happens once when the Dart object is created. If they reuse it then we avoid extra conversion and just do straight mem-copies on subsequent uses in DLs.

Also, does access to the Dart arrays require this code to execute on some pre-determined isolate? We can pass a DlColor around between isolates, but Dart data is more particular.

Reply:
We need to copy the data from the dart typed data array synchronously, we can't hold onto it outside of whatever specific ffi scope due to GC and stuff. So we're going to copy the data out into the DL, but I believe this code is first copying the data into a temporary vector which shouldn't be necessary.

Reply:
> All this code is doing is passing the data into the DLBuilder so it can be copied into the display list. Can we copy into the final destination instead of allocating an intermediary?

Yes, we could do that if we changed the signatures of the `DlColorSource` methods to take in `tonic` datatypes.

Reply:
We could also use the iterator pattern to abstract away the tonic container, that way `DlColorSource` wouldn't know it's reading from a tonic container.

Reply:
please don't use C++ iterators :) . You could safely cast the tonic data to Scalar[] and then have the DLColor constructor accept it?

Reply:
> please don't use C++ iterators :)
Hey, just because C++ iterators are wonky doesn't mean the whole pattern is bad, haha.

> You could safely cast the tonic data to Scalar[] and then have the DLColor constructor accept it?

Yea, it looks like we could do that, not a cast but there is the method `tonic::TypedList::data()`.  I'll give that a shot to see if it's something easy we can do here or is better off in a separate PR.

Reply:
Please do this in this PR and not later.

Reply:
Okay, this is going to work: https://github.com/gaaclarke/engine/commit/5ae03b571683f7c23b979391bc04c28f58ccd9b8

I'd recommend landing that change in a different PR though so that we can land the integration test before refactoring display lists, prioritizing getting everything working before making it faster.  The above diff breaks tests in about 50 places, we'll have to do that for every gradient type.  I could overload the DlColorSource methods to take in DlScalar arrays or DlColor arrays to make that easier to manage.

I also want to fix `DlColorSource::stop_count_` since it doesn't actually mean `stop_count_` but color count.

Reply:
Why does that need to land in a separate PR?

Reply:
> Why does that need to land in a separate PR?

So that we can land the integration test before landing a more invasive change.  Simple change(engine) -> integration test (framework) -> invasive change (engine).

Reply:
The logic that I'm asking you to add does not need an intergration test to cover it (in fact it may already be tested)

Reply:
> We need to copy the data from the dart typed data array synchronously, we can't hold onto it outside of whatever specific ffi scope due to GC and stuff. So we're going to copy the data out into the DL, but I believe this code is first copying the data into a temporary vector which shouldn't be necessary.

This is outside the scope of filling the DL. This is a standalone object that needs to hold the data. At some point in its future lifetime (perhaps multiple times), it will be involved in a call to fill a DL, but we don't know when in the future.

ui.Canvas call has a ui.Paint which has a ui.Shader which might be a Gradient that might have been initialized from an array of ui.Colors.

So, the gradient is constructed from the array and makes a DlColorSource. It then could just sit there forever with that value. At some point some Dart code puts it into a ui.Paint and again it could just get stashed somewhere for an indeterminant time. Then at some point that ui.Paint gets used in a ui.Canvas call - and then we copy into the DL...

Reply:
> please don't use C++ iterators :) . You could safely cast the tonic data to Scalar[] and then have the DLColor constructor accept it?

Scalar[]+offset, or I guess we could just do pointer math.

Reply:
I'd prefer we use a named factory so that it is clear what order the scalars are supposed to be in the array.

Reply:
I think I'm gradually absorbing the concern here. Talk of "DL" has me thinking of the display list object itself with recorded operations, but I think the concern here is that DlFooGradientColorSource embeds the data into a contiguous chunk and we are converting from one array to a temp vector of DlColor objects and then into the data co-located with the eventual DlColorSource object.

So, yes, it would be nice if we could copy directly from the constructor/factory argument into the pod allocation without first creating a vector of DlColor objects. I'd be happy to see that added as a future optimization and just stick with making some sort of list of DlColor objects for now to get this capability implemented in its simplest fashion.

We could also consider having the storage for these gradient objects be just an array of floats and a separate common CS. The gradient objects could still get you a single color in a getter by marrying the right 4 floats and the CS into a DlColor object, but there would also be accessors to the common CS value and the array of component float data which the backend might use more efficiently. It would also save on storage size.

Reply:
Also, I will be coming through these classes to de-skia-fy them shortly, so I can implement such optimizations as I do that work. That work should all fall under "should have no impact at all on output" changes and so would hopefully pass up the chain without regressions.

Reply:
> I'd prefer we use a named factory so that it is clear what order the scalars are supposed to be in the array.

That's what I'm doing I'm refactoring DlColorSource to use the Builder pattern.  That way we can pass in DlColor arrays or float arrays.

> The logic that I'm asking you to add does not need an intergration test to cover it (in fact it may already be tested)

This PR is adding new functionality that is only tested in an integration test that has to live in another repo.  This PR implements the feature while working within the bounds of the current API.  I agree with your feedback and it will be addressed.  That's why we went through the whole process of brainstorming a solution.  However this PR does not add any new heap allocations, they were already there, removing them is going to be a bit of a refactor, one who would be served by having a test as a backstop to make sure we don't break the thing we are actually trying to implement.

We can withhold the approval of this PR until you see the refactor PR if you want.  I don't understand why I'm not being trusted here.

Reply:
Removing an unnecessary heap allocation does not require an integration test

Reply:
Moving to the functions to DlScalar* is not going to work because there are 2 clients of `DlLinearGradientColorSource::MakeLinear`.  In one case we have DlScalar*, in another we have DlColor*.  Moving to DlScalar would require us to do heap allocations when converting in the other route.  The other client is `DlColorSource::shared()`, I tried removing it and that runs deep coming from DlAttribute.

The only viable solution is to use the iterator pattern if we want to eliminate heap allocations in both cases.  That would require moving more logic to templates or a dynamic dispatch, which would possibly eliminate any benefit of removing the tiny heap allocation.

I think the best bet is to eliminate the heap in most cases but have an overflow for gradients with large amounts of colors.